### PR TITLE
[FLINK-10687] [table] Make flink-formats Scala-free

### PR DESCRIPTION
**Pull request by twalthr**
## What is the purpose of the change

This PR makes the `flink-formats` module Scala free by introducing a `flink-table-common` module that is implemented in Java. This module contains all classes that are required across different Maven modules.

## Brief change log

The following classes have been migrated to Java:

```
./src/test/java/org/apache/flink/table/descriptors/DescriptorTestBase.java
./src/test/java/org/apache/flink/table/factories/TableFormatFactoryBaseTest.java
./src/main/java/org/apache/flink/table/utils/TypeStringUtils.java
./src/main/java/org/apache/flink/table/utils/EncodingUtils.java
./src/main/java/org/apache/flink/table/descriptors/Descriptor.java
./src/main/java/org/apache/flink/table/descriptors/DescriptorBase.java
./src/main/java/org/apache/flink/table/descriptors/ConnectorDescriptorValidator.java
./src/main/java/org/apache/flink/table/descriptors/FormatDescriptor.java
./src/main/java/org/apache/flink/table/descriptors/DescriptorValidator.java
./src/main/java/org/apache/flink/table/descriptors/FormatDescriptorValidator.java
./src/main/java/org/apache/flink/table/descriptors/DescriptorProperties.java
./src/main/java/org/apache/flink/table/descriptors/ConnectorDescriptor.java
./src/main/java/org/apache/flink/table/factories/TableFactory.java
./src/main/java/org/apache/flink/table/factories/TableFormatFactory.java
./src/main/java/org/apache/flink/table/factories/DeserializationSchemaFactory.java
./src/main/java/org/apache/flink/table/factories/SerializationSchemaFactory.java
./src/main/java/org/apache/flink/table/factories/TableFormatFactoryBase.java
./src/main/java/org/apache/flink/table/api/ValidationException.java
./src/main/java/org/apache/flink/table/api/TableException.java
./src/main/java/org/apache/flink/table/api/TableSchema.java
```

Additionally, all classes in this module have been annotated with `@Internal` and `@PublicEvolving` accordingly.

Since the methods in `Descriptor` were declared with `private[flink]` visibility, they have been migrated to a new `toProperties()` method that is public.

## Verifying this change

All existing tests are still successful.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): yes
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? JavaDocs

### CODE REVIEW

File: flink-formats/flink-avro/pom.xml [hunk 1]
```diff
@@ -64,16 +64,30 @@ under the License.
...
+			<artifactId>flink-table-common</artifactId>
 			<version>${project.version}</version>
 			<scope>provided</scope>
-			<!-- Projects depending on this project, won't depend on flink-table. -->
-			<optional>true</optional>
 		</dependency>
 
 		<!-- test dependencies -->
 
+		<dependency>
```
Comment (aljoscha) [inline flink-formats/flink-avro/pom.xml]: what's the test dependency used for?

File: flink-formats/flink-json/src/main/java/org/apache/flink/formats/json/JsonRowFormatFactory.java [hunk 2]
```diff
@@ -23,51 +23,39 @@
...
 	}
 
 	@Override
-	public boolean supportsSchemaDerivation() {
-		return true;
-	}
-
-	@Override
-	public List<String> supportedProperties() {
+	protected List<String> supportedFormatProperties() {
```
Comment (aljoscha) [inline flink-formats/flink-json/src/main/java/org/apache/flink/formats/json/JsonRowFormatFactory.java]: Why is this removed, and the next part where `FORMAT_DERIVE_SCHEMA` is removed?

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/descriptors/SchemaValidatorTest.scala [hunk 3]
```diff
@@ -150,15 +142,6 @@ class SchemaValidatorTest {
     assertEquals(
       expectedMapping,
       SchemaValidator.deriveFieldMapping(props, Optional.of(inputSchema.toRowType)))
-
```
Comment (aljoscha) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/descriptors/SchemaValidatorTest.scala]: Why is this part of the test removed?

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/descriptors/SchemaValidatorTest.scala [hunk 4]
```diff
@@ -69,14 +69,6 @@ class SchemaValidatorTest {
...
```
Comment (aljoscha) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/descriptors/SchemaValidatorTest.scala]: Why is this part of the test removed?

File: flink-libraries/flink-table-common/pom.xml [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+	<dependencies>
+		<dependency>
+			<groupId>org.apache.flink</groupId>
+			<artifactId>flink-core</artifactId>
+			<version>1.7-SNAPSHOT</version>
+		</dependency>
+	</dependencies>
+
+	<build>
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-libraries/flink-table-common/pom.xml]: Yes, maybe not that much right now but in the future. Currently, other modules can use `org.apache.flink.table.descriptors.DescriptorTestBase`. In the future it is likely that more classes will be added as `flink-table-common` is intended to be shared across modules.

File: flink-libraries/flink-table-common/src/main/java/org/apache/flink/table/api/TableSchema.java [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+				"List of duplicate fields: " + duplicateNames.toString() + "\n" +
+				"List of all fields: " + Arrays.toString(fieldNames));
+		}
+	}
+
+	/**
+	 * Returns a deep copy of the table schema.
+	 */
+	public TableSchema copy() {
+		return new TableSchema(fieldNames.clone(), fieldTypes.clone());
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-libraries/flink-table-common/src/main/java/org/apache/flink/table/api/TableSchema.java]: I don't think that we needed it. `TypeInformation` should be immutable. The current Scala implementation is also like this.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/typeutils/TypeStringUtilsTest.scala [hunk 7]
```diff
@@ -103,13 +104,18 @@ class TypeStringUtilsTest {
 
     // test escaping
     assertTrue(
-      TypeStringUtils.readTypeInfo("ROW<\"he         \\nllo\" DECIMAL, world TINYINT>")
+      TypeStringUtils.readTypeInfo("ROW<`he         \nllo` DECIMAL, world TINYINT>")
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-libraries/flink-table/src/test/scala/org/apache/flink/table/typeutils/TypeStringUtilsTest.scala]: The handling of escaping was simply wrong in this class. Similar efforts have been done recently for https://issues.apache.org/jira/browse/FLINK-10281. Backticks (or double backticks) are used in SQL and also other vendors such as Hive.

File: flink-formats/flink-avro/pom.xml (hunk 1 above)
Comment (twalthr) [reply flink-formats/flink-avro/pom.xml]: This dependency is necessary for `org.apache.flink.table.descriptors.AvroTest`.

File: flink-formats/flink-json/src/main/java/org/apache/flink/formats/json/JsonRowFormatFactory.java (hunk 2 above)
Comment (twalthr) [reply flink-formats/flink-json/src/main/java/org/apache/flink/formats/json/JsonRowFormatFactory.java]: I introduced a base class that contains the schema derivation logic and reduces code duplication.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/descriptors/SchemaValidatorTest.scala (hunk 3 above)
Comment (twalthr) [reply flink-libraries/flink-table/src/test/scala/org/apache/flink/table/descriptors/SchemaValidatorTest.scala]: I readded the test to `org.apache.flink.table.factories.TableFormatFactoryBaseTest#testSchemaDerivation`.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/descriptors/SchemaValidatorTest.scala (hunk 4 above)
Comment (twalthr) [reply flink-libraries/flink-table/src/test/scala/org/apache/flink/table/descriptors/SchemaValidatorTest.scala]: I moved the test to `org.apache.flink.table.factories.TableFormatFactoryBaseTest#testSchemaDerivation`.

**Comment by aljoscha** [comment]
I did review everything except the new parser. Are there any tests that test more complex things, like nested types, deep nesting, and things?

Otherwise this looks good. 👍

**Comment by twalthr** [comment]
Thanks for the review @aljoscha. I added more tests and improvements to the type parser.

**Comment by bowenli86** [comment]
I only had a detailed look at classes in `flink-table-common`. Big PR while changes are mostly renaming and migrating without introducing new feature. Look very good to me!

**Comment by twalthr** [comment]
Thanks everyone. I will merge this now...
